nni/compression/pytorch/quantization_speedup/integrated_tensorrt.py

```diff
@@ -0,0 +1,337 @@
+    8: trt.float32,
+    16: trt.float16,
+    32: trt.float32
+}
+
+def valid_config(config=None):
+    """
+    This function validates the bit setting configuration
+    """
+    if config is None:
+        return
+    support_bit = [8, 16, 32]
+    for name in config.keys():
+        if 'weight_bit' in config[name]:
+            w_bit = config[name]['weight_bit']
+            assert w_bit in support_bit, "weight bit should be 8, 16, 32"
+        if 'activation_bit' in config[name]:
+            a_bit = config[name]['activation_bit']
+            assert a_bit in support_bit, "activation bit should be 8, 16, 32"
+
+def build_engine(model_file, config=None, extra_layer_bit=32, strict_datatype=False, calib=None):
+    """
+    This function builds an engine from an onnx model with calibration process.
+    Parameters
+    ----------
+    model_file : str
+        The path of onnx model
+    config : dict
+        Config recording bit number and name of layers
+    extra_layer_bit : int
+        Other layers which are not in config will be quantized to corresponding bit number
+    strict_datatype : bool
+        Whether constrain layer bit to the number given in config or not. If true, all the layer
+        will be set to given bit strictly. Otherwise, these layers will be set automatically by
+        tensorrt
+    calib : numpy array
+        The data using to calibrate quantization model
+    Returns
+    -------
+    engine : tensorrt.ICudaEngine
+        An ICudaEngine for executing inference on a built network
+    """
+    with trt.Builder(TRT_LOGGER) as builder, builder.create_network(common.EXPLICIT_BATCH) as network, \
+        trt.OnnxParser(network, TRT_LOGGER) as parser:
+        # Attention that, builder should be set to 1 because of the implementation of allocate_buffer
+        builder.max_batch_size = 1
+        builder.max_workspace_size = common.GiB(4)
+
+        if extra_layer_bit == 32 and config is None:
+            pass
+        elif extra_layer_bit == 16 and config is None:
+            builder.fp16_mode = True
+        elif extra_layer_bit == 8 and config is None:
+            # entire model in 8bit mode
+            builder.int8_mode = True
+        else:
+            builder.int8_mode = True
+            builder.fp16_mode = True
+            builder.strict_type_constraints = strict_datatype
+
+        # Parse onnx model
+        with open(model_file, 'rb') as model:
+            if not parser.parse(model.read()):
+                logger.error('ERROR: Fail to parse the ONNX file.')
+                for error in range(parser.num_errors):
+                    logger.error(parser.get_error(error))
+                return None
+
+        if calib is not None:
+            builder.int8_calibrator = calib
+            # This design may not be correct if output more than one
+            for i in range(network.num_layers):
+                if config is None:
+                    break
+                valid_config(config)
```

Comment:
`valid_config` should be called many times?

Reply:
Have modified.